### PR TITLE
FEAT Introduce StreamingMetricFrame to accumulate batches

### LINKED ISSUE #655: Support fort streaming metrics
#### Is your feature request related to a problem? Please describe.

I am using fairlearn to evaluate deep learning models. In this context, we often compute metrics per batch due to the large size of the dataset. In addition, some tasks such as semantic segmentation are so expensive that we can't hold all the predictions in memory.

#### Describe the solution you'd like
I would like if `MetricFrame` would accept an `update` method where we can add more data. For the first version, we can simply add more data to the internal representation of `ytrue, ypred, sensitive_attribute`. But then I would like it if we could keep a confusion matrix and the `update` method would update this confusion matrix.

#### Describe alternatives you've considered, if relevant

I made a wrapper around Fairlearn metrics that keep tracks of `ytrue, ypred, sensitive_attribute` and then call the underlying metric when needed.

#### Additional context

Example of a metric in Pytorch Lightning, a fairly popular DL framework.
https://pytorch-lightning.readthedocs.io/en/stable/metrics.html#implementing-a-metric

Keras has a similar API

https://keras.io/api/metrics/#as-subclasses-of-metric-stateful

### PR DESCRIPTION
**Pull request by Dref360**
Introduce StreamingMetricFrame to accumulate batches and enable batch evaluation.

This version focuses on maintainability and is much easier to work with than the previous one. Most checks are now done by MetricFrame.

Fixes #655 

### Notes

In the previous PR (#670), we mentioned that converting everything to pandas would be valuable.
I think this would be useful, but this approach gets complicated with nested dict found in `sample_params`. I am open to propositions on how to fix that and maybe get rid of `StreamingMetricFrame._concat_batches`.

### CODE REVIEW

File: fairlearn/metrics/_streaming_metric_frame.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            self._control_features = self._concat_if_not_none(self._control_features,
+                                                              control_features)
+        elif self._control_features is not None:
+            raise ValueError(_CF_BAD_STATE_ERR)
+
+        if sample_params is not None:
+            self._sample_params = self._concat_if_not_none(self._sample_params,
+                                                           sample_params)
+
+    def get_metric(self) -> MetricFrame:
```
Comment (riedgar-ms) [inline fairlearn/metrics/_streaming_metric_frame.py:124]: Perhaps `evaluate_metrics()` ? @MiroDudik @hildeweerts @adrinjalali ?

File: fairlearn/metrics/_streaming_metric_frame.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            a nested dictionary, with the first set of string keys identifying the
+            metric function name, with the values being the string-to-array-like dictionaries.
+
+            **Note** the types returned by members of the class vary based on whether
+            control features are present.
+        """
+        check_consistent_length(y_true, y_pred)
+        self._y_true = self._concat_if_not_none(self._y_true, y_true)
+        self._y_pred = self._concat_if_not_none(self._y_pred, y_pred)
+        self._sensitive_features = self._concat_if_not_none(self._sensitive_features,
```
Comment (riedgar-ms) [inline fairlearn/metrics/_streaming_metric_frame.py:112]: Should sensitive features be checked for consistent length?

File: fairlearn/metrics/_streaming_metric_frame.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                                                            sensitive_features)
+        if control_features is not None:
+            self._control_features = self._concat_if_not_none(self._control_features,
+                                                              control_features)
+        elif self._control_features is not None:
+            raise ValueError(_CF_BAD_STATE_ERR)
+
+        if sample_params is not None:
+            self._sample_params = self._concat_if_not_none(self._sample_params,
+                                                           sample_params)
```
Comment (riedgar-ms) [inline fairlearn/metrics/_streaming_metric_frame.py:122]: Shouldn't this have an else clause like the control features? And a consistent length check?

File: fairlearn/metrics/_streaming_metric_frame.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            **Note** the types returned by members of the class vary based on whether
+            control features are present.
+        """
+        check_consistent_length(y_true, y_pred)
+        self._y_true = self._concat_if_not_none(self._y_true, y_true)
+        self._y_pred = self._concat_if_not_none(self._y_pred, y_pred)
+        self._sensitive_features = self._concat_if_not_none(self._sensitive_features,
+                                                            sensitive_features)
+        if control_features is not None:
+            self._control_features = self._concat_if_not_none(self._control_features,
```
Comment (riedgar-ms) [inline fairlearn/metrics/_streaming_metric_frame.py:115]: Consistent length?

**Review by riedgar-ms — COMMENTED** [review]
Thank you for trying again at this!

File: fairlearn/metrics/_streaming_metric_frame.py (hunk 1 above)
Comment (adrinjalali) [reply fairlearn/metrics/_streaming_metric_frame.py:124]: that's a good question. Somehow I was thinking of this being sth that I could add data to, and then call `.evaluate()` which either returns nothing or `self`, and it would have all the features like `MetrifFrame`. If we don't want to go down that path, I think `get_metricframe` would be what this does?

**Comment by hildeweerts** [comment]
At a first glance it seems like @Dref360 has processed al requested changes from @riedgar-ms. Is there anything else that's blocking this PR from getting merged?

Tagging @fairlearn/fairlearn-maintainers

File: fairlearn/metrics/_streaming_metric_frame.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+            alongside the ``y_true`` and ``y_pred`` arrays, and passed to the metric function.
+            If there are multiple metric functions (passed as a dictionary), then this is
+            a nested dictionary, with the first set of string keys identifying the
+            metric function name, with the values being the string-to-array-like dictionaries.
+
+            **Note** the types returned by members of the class vary based on whether
+            control features are present.
+        """
+        check_consistent_length(y_true, y_pred)
+        check_consistent_length(y_true, sensitive_features)
```
Comment (adrinjalali) [inline fairlearn/metrics/_streaming_metric_frame.py:110]: you could merge these calls and pass the three objects

**Review by adrinjalali — COMMENTED** [review]
Otherwise LGTM, except that I'm not happy that we don't have codecov and I can't check if tests cover the new code. I guess I should set it up.
